### PR TITLE
Pbi 1678

### PR DESCRIPTION
**Pull request by aeng7**
-added recovery mechanisms for the load, start, and stop events of vpaid creatives

### CODE REVIEW

**Comment by aeng7** [comment]
Resolving conflicts.

File: js/ad_manager_vast.js [hunk 1]
```diff
@@ -641,9 +659,51 @@ OO.Ads.manager(function(_, $) {
...
 
+    var _clearVpaidTimeouts = function() {
+      clearTimeout(vpaidIframeLoadedTimeout);
+      vpaidIframeLoadedTimeout = null;
+
+      clearTimeout(vpaidAdLoadedTimeout);
+      vpaidAdLoadedTimeout = null;
+
+      clearTimeout(vpaidAdStartedTimeout);
+      vpaidAdStartedTimeout = null;
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: do we not care about clearing out the stopped timeout?

File: js/ad_manager_vast.js [hunk 2]
```diff
@@ -667,8 +727,28 @@ OO.Ads.manager(function(_, $) {
...
+
+          if (typeof pbMetadata.vpaidTimeout.loaded === "number") {
+            this.VPAID_AD_LOADED_TIMEOUT = pbMetadata.vpaidTimeout.loaded * 1000;
+          }
+
+          if (typeof pbMetadata.vpaidTimeout.started === "number") {
+            this.VPAID_AD_STARTED_TIMEOUT = pbMetadata.vpaidTimeout.started * 1000;
+          }
+
+          if (typeof pbMetadata.vpaidTimeout.stopped === "number") {
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: Two things: Is 0 valid --> Should we do if( timeoutVar && _.isNumber(timeoutVar))?

Are negative values valid?

File: test/unit-tests/vast_test.js [hunk 3]
```diff
@@ -1787,6 +1787,33 @@ describe('ad_manager_vast', function() {
...
+    vastAdManager.loadMetadata({"vpaidTimeout":{
+      "iframe":9,
+      "loaded":10,
+      "started":11,
+      "stopped":12
+    }}, {}, content);
+    expect(vastAdManager.VPAID_AD_IFRAME_TIMEOUT).to.be(9000);
+    expect(vastAdManager.VPAID_AD_LOADED_TIMEOUT).to.be(10000);
+    expect(vastAdManager.VPAID_AD_STARTED_TIMEOUT).to.be(11000);
+    expect(vastAdManager.VPAID_AD_STOPPED_TIMEOUT).to.be(12000);
```
Comment (pilievOoyala) [inline test/unit-tests/vast_test.js]: not for this checkin, but I think Angie knows a way to simulate time. So we can actually check if the timeouts work.

File: js/ad_manager_vast.js (hunk 1 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Woops! Awesome catch.

**Comment by pilievOoyala** [comment]
Looks good!

**Pull request merged**